drivers/spi/spi_mspm0.c

```diff
@@ -0,0 +1,353 @@
+	const struct mspm0_sys_clock *clock_subsys;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+struct spi_mspm0_data {
+	struct spi_context spi_ctx;
+	struct k_sem spi_idle;
+};
+
+static void spi_mspm0_isr(const struct device *dev)
+{
+	const struct spi_mspm0_config *config = dev->config;
+	struct spi_mspm0_data *data = dev->data;
+
+	if (DL_SPI_getPendingInterrupt(config->spi_base) == DL_SPI_IIDX_IDLE) {
+		DL_SPI_disableInterrupt(config->spi_base, DL_SPI_INTERRUPT_IDLE);
+		k_sem_give(&data->spi_idle);
+	}
+}
+
+static int spi_mspm0_configure(const struct device *dev, const struct spi_config *spi_cfg,
+			       uint8_t dfs)
+{
+	const struct spi_mspm0_config *const config = dev->config;
+	struct spi_mspm0_data *const data = dev->data;
+	const struct device *clk_dev = DEVICE_DT_GET(DT_NODELABEL(ckm));
+
+	uint32_t clock_rate;
+	uint16_t clock_scr;
+	int ret;
+
+	if (spi_context_configured(&data->spi_ctx, spi_cfg)) {
+		/* This configuration is already in use */
+		return 0;
+	}
+
+	/* Only master mode is supported */
+	if (SPI_OP_MODE_GET(spi_cfg->operation) != SPI_OP_MODE_MASTER) {
+		return -ENOTSUP;
+	}
+
+	ret = clock_control_get_rate(clk_dev, (struct mspm0_sys_clock *)config->clock_subsys,
+				     &clock_rate);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (spi_cfg->frequency > (clock_rate / 2)) {
+		return -EINVAL;
+	}
+
+	/* See DL_SPI_setBitRateSerialClockDivider for details */
+	clock_scr = (clock_rate / (2 * spi_cfg->frequency)) - 1;
+	if (!IN_RANGE(clock_scr, MSPM0_SPI_SCR_MIN, MSPM0_SPI_SCR_MAX)) {
+		return -EINVAL;
+	}
+
+	const DL_SPI_Config dl_spi_cfg = {
+		.parity = DL_SPI_PARITY_NONE,			/* Currently unused in zephyr */
+		.chipSelectPin = DL_SPI_CHIP_SELECT_NONE,	/* spi_context controls the CS */
+		.mode = SPI_MODE(spi_cfg->operation),
+		.bitOrder = BIT_ORDER_MODE(spi_cfg->operation),
+		.dataSize = DATA_SIZE_MODE(spi_cfg->operation),
+		.frameFormat = FRAME_FORMAT_MODE(spi_cfg->operation),
+	};
+
+	/* Peripheral should be disabled before applying a new configuration */
+	DL_SPI_disable(config->spi_base);
+
+	DL_SPI_init(config->spi_base, (DL_SPI_Config *)&dl_spi_cfg);
+	DL_SPI_setBitRateSerialClockDivider(config->spi_base, (uint32_t)clock_scr);
+
+	if (dfs > SPI_DFS_16BIT) {
+		DL_SPI_enablePacking(config->spi_base);
+	} else {
+		DL_SPI_disablePacking(config->spi_base);
+	}
+
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_LOOP) {
+		DL_SPI_enableLoopbackMode(config->spi_base);
+	} else {
+		DL_SPI_disableLoopbackMode(config->spi_base);
+	}
+
+	DL_SPI_enable(config->spi_base);
+
+	/* Cache SPI config for reuse, required by spi_context owner */
+	data->spi_ctx.config = spi_cfg;
+
+	return 0;
+}
+
+static void spi_mspm0_frame_tx(const struct device *dev, uint8_t dfs)
+{
+	const struct spi_mspm0_config *config = dev->config;
+	struct spi_mspm0_data *data = dev->data;
+
+	/* Transmit dummy frame when no TX data is provided */
+	uint32_t tx_frame = 0;
+
+	k_sem_reset(&data->spi_idle);
```

Comment:
This is not protecting anything and MULTITHREADING will fail. Anyways, move to spi_context